### ISSUE TITLE
Unable to use IteratorDatasets

### ISSUE THREAD

**Issue by sagostinho-argo** [determined-ai/determined#3660]
Hey everyone. I'm currently trying to use a PyTorch iterator dataset, where I'm responsible for handling sharding, shuffling, etc...

According to the source code
https://github.com/determined-ai/determined/blob/f3f15e8dcd4de1f745b762f2edba5b2568416d59/harness/determined/pytorch/_data.py#L105-L108

and respective docs
https://docs.determined.ai/0.17.2/training-apis/api-pytorch-samplers.html
> Normally, using det.pytorch.DataLoader is required and handles all of the below details without any special effort on your part (see [Data Loading](https://docs.determined.ai/0.17.2/training-apis/api-pytorch.html#pytorch-data-loading)). When det.pytorch.DataLoader() is not suitable (especially in the case of **IterableDatasets**), you may disable this requirement by calling [context.experimental.disable_dataset_reproducibility_checks()](https://docs.determined.ai/0.17.2/training-apis/api-pytorch.html#determined.pytorch.PyTorchExperimentalContext.disable_dataset_reproducibility_checks) in your Trial’s __init__() method. Then you may choose to follow the below guidelines for ensuring dataset reproducibility on your own.

this should be possible. However, you seem to breach this premise in
https://github.com/determined-ai/determined/blob/f3f15e8dcd4de1f745b762f2edba5b2568416d59/harness/determined/pytorch/_pytorch_trial.py#L360-L361

You're pulling from the iterator using `next`, but you're not catching `StopIteration` exceptions. 

Similar thing in your master branch
https://github.com/determined-ai/determined/blob/754964184840a8fab698d762f0567598bd248893/harness/determined/pytorch/_pytorch_trial.py#L378-L379

Am I missing something obvious here?

Thank you for the help.

**Comment by rb-determined-ai** [comment]
I'm glad you are taking this feature for a spin!  (full disclosure: I wrote it)

We don't catch `StopIteration` because our training loop requires that your training iterator is infinite.  That should be covered in the sampler docs.  We could definitely detect the StopIteration and raise a better error to explain that.

Also, do you have other feedback on the docs or the feature?

**Comment by sagostinho-argo** [comment]
Thank you for chiming in. 

> We don't catch StopIteration because our training loop requires that your training iterator is infinite. That should be covered in the sampler docs. We could definitely detect the StopIteration and raise a better error to explain that.

I clearly missed this. The concept of "repeating" applied to a dataset is tensorflow specific, so coming from a pytorch background, I just assumed it was some sampler specific operation.  I didn't catch the intended meaning. I'll dig through the docs again keeping in mind this new info. 

> Also, do you have other feedback on the docs or the feature?

I personally only assumed that an iterator's dataset only restriction was that it did not allow random access. I had no idea there that it common to assume that it is "infinite". I mention this because I work with certain datasets that are clearly fine to hold in storage and are "finite", but where doing random access is just prohibitively expensive. Parsing properly the `StopIteration` exception would allow to have a proper notions of epochs.

Out of curiosity, is there anything technical that is not possible to perform correctly if we allow the dataset to end  (and iterate again as a new epoch), e.g., some sampling or shuffling limitation?

**Comment by rb-determined-ai** [comment]
> I had no idea there that it common to assume that it is "infinite".

I would not say this is common to assume, I would say it is a quirk of our PyTorchTrial training loop.  I think the more intuitive behavior is, as you say, to allow the dataset to end and iterate on a new epoch.  This is actually fairly tricky in distributed training because at the end of a dataset, all the workers have to coordinate who is still training, and then the distributed training framework has to be able to allreduce the gradient updates from arbitrary number of workers, and the non-participating workers still need to update their model weights according to updates calculated by the other workers.

So we would like to let data loaders run out and then just iterate through them again, but at this time we do not support that.
